python/pyfury/_registry.py

```diff
@@ -0,0 +1,627 @@
+        for itemsize, ftype, typeid in PyArraySerializer.typecode_dict.values():
+            register(
+                ftype,
+                type_id=typeid,
+                serializer=PyArraySerializer(self.fury, ftype, typeid),
+            )
+        register(
+            array.array, type_id=DYNAMIC_TYPE_ID, serializer=DynamicPyArraySerializer
+        )
+        if np:
+            # overwrite pyarray  with same type id.
+            # if pyarray are needed, one must annotate that value with XXXArrayType
+            # as a field of a struct.
+            for dtype, (
+                itemsize,
+                format,
+                ftype,
+                typeid,
+            ) in Numpy1DArraySerializer.dtypes_dict.items():
+                register(
+                    ftype,
+                    type_id=typeid,
+                    serializer=Numpy1DArraySerializer(self.fury, ftype, dtype),
+                )
+            register(np.ndarray, type_id=DYNAMIC_TYPE_ID, serializer=NDArraySerializer)
+        register(list, type_id=TypeId.LIST, serializer=ListSerializer)
+        register(set, type_id=TypeId.SET, serializer=SetSerializer)
+        register(dict, type_id=TypeId.MAP, serializer=MapSerializer)
+        try:
+            import pyarrow as pa
+            from pyfury.format.serializer import (
+                ArrowRecordBatchSerializer,
+                ArrowTableSerializer,
+            )
+
+            register(
+                pa.RecordBatch,
+                type_id=TypeId.ARROW_RECORD_BATCH,
+                serializer=ArrowRecordBatchSerializer,
+            )
+            register(
+                pa.Table, type_id=TypeId.ARROW_TABLE, serializer=ArrowTableSerializer
+            )
+        except Exception:
+            pass
+
+    def register_type(
+        self,
+        cls: Union[type, TypeVar],
+        *,
+        type_id: int = None,
+        namespace: str = None,
+        typename: str = None,
+        serializer=None,
+    ):
+        return self._register_type(
+            cls,
+            type_id=type_id,
+            namespace=namespace,
+            typename=typename,
+            serializer=serializer,
+        )
+
+    def _register_type(
+        self,
+        cls: Union[type, TypeVar],
+        *,
+        type_id: int = None,
+        namespace: str = None,
+        typename: str = None,
+        serializer=None,
+        internal=False,
+    ):
+        """Register class with given type id or typename. If typename is not None, it will be used for
+        cross-language serialization."""
+        if serializer is not None and not isinstance(serializer, Serializer):
+            try:
+                serializer = serializer(self.fury, cls)
+            except BaseException:
+                try:
+                    serializer = serializer(self.fury)
+                except BaseException:
+                    serializer = serializer()
+        n_params = len({typename, type_id, None}) - 1
```

Comment:
```suggestion
        n_params = len((typename, type_id, None)) - 1
```
In CPython, both `tuple` and `list` use a freelists to speed up allocation and deallocation, whereas `set` and `dict` do not have this. Therefore, using a tuple or list is more appropriate if you need a temporary structure.

Reply:
We need set here, because we are deduplicating the arguments to check `n_params`. And this is not in critical path, the performance is not important

Reply:
lgtm